data/collector_test.go

```diff
@@ -65,101 +67,120 @@ func TestNewCollector(t *testing.T) {
 		},
 		{
-			name:        "Sleep based binary writer.",
-			captureFunc: dummyBinaryCapturer,
-			params: CollectorParams{
-				ComponentName: "testComponent",
-				Interval:      sleepInterval,
-				MethodParams:  map[string]*anypb.Any{"name": fakeVal},
-				QueueSize:     queueSize,
-				BufferSize:    bufferSize,
-				Logger:        l,
-			},
-			wait:           sleepInterval*time.Duration(2) + sleepInterval/time.Duration(2),
+			name:           "Sleep based binary writer.",
+			captureFunc:    binaryCapturer,
+			interval:       sleepInterval,
 			expectReadings: 2,
 			expFiles:       2,
 		},
 	}
 
 	for _, tc := range tests {
 		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(time.Second))
+			defer cancel()
 			tmpDir := t.TempDir()
 			md := v1.DataCaptureMetadata{}
-			target := datacapture.NewBuffer(tmpDir, &md)
-			test.That(t, target, test.ShouldNotBeNil)
+			tgt := datacapture.NewBuffer(tmpDir, &md)
+			test.That(t, tgt, test.ShouldNotBeNil)
+			wrote := make(chan struct{})
+			target := &signalingBuffer{
+				bw:    tgt,
+				wrote: wrote,
+			}
 
-			tc.params.Target = target
-			c, err := NewCollector(tc.captureFunc, tc.params)
+			mockClock := clock.NewMock()
+			params.Interval = tc.interval
+			params.Target = target
+			params.Clock = mockClock
+			c, err := NewCollector(tc.captureFunc, params)
 			test.That(t, err, test.ShouldBeNil)
 			c.Collect()
+			// We need to avoid adding time until after the the underlying goroutine has started sleeping.
+			// If we add time before that point, data will never be captured, because time will never be greater than
+			// the initially calculated time.
+			// Sleeping for 1ms is a hacky way to ensure that we don't encounter this situation. It gives 1ms
+			// for those few sequential lines in collector.go to execute, so that that occurs before we add time below.
+			time.Sleep(time.Millisecond)
+			for i := 0; i < tc.expectReadings; i++ {
+				mockClock.Add(params.Interval)
+				select {
+				case <-ctx.Done():
+					t.Fatalf("timed out waiting for data to be written")
+				case <-wrote:
+				}
+			}
+			close(wrote)
 
-			// Verify that the data it wrote matches what we expect.
-			time.Sleep(tc.wait)
+			// If it's a sleep based collector, we need to move the clock forward one more time after calling Close.
+			// Otherwise, it will stay asleep indefinitely and Close will block forever.
+			// This loop guarantees that the clock is moved forward at least once after Close is called. After Close
+			// returns and the closed channel is closed, this loop will terminate.
+			closed := make(chan struct{})
+			sleepCollector := tc.interval < sleepCaptureCutoff
+			wg := sync.WaitGroup{}
+			if sleepCollector {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					for i := 0; i < 100; i++ {
+						select {
+						case <-closed:
+							return
+						default:
+							time.Sleep(time.Millisecond * 10)
```

Comment:
Thank you for the detailed descriptions, super helpful for both me and anyone reading this RDK codebase.
[superdupernit] but does this need to be 10ms vs a shorter time period?

Reply:
Nope! Updated to be 1ms and a loop of 1000 instead